### PR TITLE
Typed notification in iOS

### PR DESCRIPTION
**Pull request by aksswami**

### CODE REVIEW

File: _posts/2017-08-28-typed-notification-ios.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+}
+{% endhighlight %}
+
+We write this same code again and again or we create a `BaseViewController` and have these defined or maybe a `UIViewController` extension. But all of these are still not efficient as you have to dig into `notification.userInfo` dictionary to get relevant information.
+
+### Typed Notificaion
+
+![Strongly Typed]({{ site.url }}/assets/images/strongly-typed-300x210.png)
+
+We want to define notifications in such a way, where we can definitively get the data from our notification observer (strongly typed), not a userInfo dictionary. This way we will enforce swift type system to help us writing type safe and bug free code. In this process we will also redefine how we observe our notification to make our code `DRY`.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: explain DRY a little bit, may be just full form in ()

File: _posts/2017-08-28-typed-notification-ios.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+title:  "Typed Notification iOS"
+date:   2017-08-28 11:11:11
+author: amit
+categories: Technical iOS
+---
+Dealing with [NotificationCenter](https://developer.apple.com/documentation/notificationcenter) is daily task for all iOS developers. Either it's a system notification such as [KeyboardShow/KeyboardHide](https://developer.apple.com/library/content/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html) or CustomNotification to post some information app wide, we all have used Notifications.
+
+### Yeah, So. Do you have something interesting to say?
+
+Well, answer to that question is, Yes. Today We are gonna discuss notification programming in such a way, that it will become as easy as eating an apple pie. (I don't think anyone can deny it, eating an apple pie is just so satisfying, obviously if you love apple pie in first place ;p )
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: Today, we ......

File: _posts/2017-08-28-typed-notification-ios.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+We write this same code again and again or we create a `BaseViewController` and have these defined or maybe a `UIViewController` extension. But all of these are still not efficient as you have to dig into `notification.userInfo` dictionary to get relevant information.
+
+### Typed Notificaion
+
+![Strongly Typed]({{ site.url }}/assets/images/strongly-typed-300x210.png)
+
+We want to define notifications in such a way, where we can definitively get the data from our notification observer (strongly typed), not a userInfo dictionary. This way we will enforce swift type system to help us writing type safe and bug free code. In this process we will also redefine how we observe our notification to make our code `DRY`.
+
+To create typed notification we can first define an `NotificationDescriptor` struct.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: define a 'NotificationDescriptor'

File: _posts/2017-08-28-typed-notification-ios.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+struct NotificationDescriptor<A> {
+    //Name of the notification
+    let name: Notification.Name
+
+    //Method to convert notification.userInfo to desired type
+    let convert: (Notification) -> A
+}
+{% endhighlight %}
+
+With help of Swift's Generic type we can use above Struct for all type of notification and describe our notifications using this.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: all type of notifications ....

File: _posts/2017-08-28-typed-notification-ios.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+Next we can extend `NotificationCenter` to provide us an convenience method over default `addObserver` method.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: a convenience ....

File: _posts/2017-08-28-typed-notification-ios.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        self.center = center
+    }
+
+    deinit {
+        center.removeObserver(token)
+    }
+}
+{% endhighlight %}
+
+And redefine out `addObserver` method this way:-
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: redefine our ...

File: _posts/2017-08-28-typed-notification-ios.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        return Token(token: t, center: self)
+    }
+}
+{% endhighlight %}
+
+We can store the `Token` returned by above function as a ViewController property, and as soon as ViewController's `deinit` gets called, `Token` will also gets destroyed and notification gets deregistered.
+
+### How to use it
+
+We will define KeyboardShow and Keyboard hide notification using our `NotificationDescriptor`.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: keyboard show and keyboard hide notifications .....

File: _posts/2017-08-28-typed-notification-ios.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    }
+}
+{% endhighlight %}
+
+`KeboardShowDescriptor` and `KeyboardHideDescriptor` will become
+
+{% highlight swift %}
+struct SystemNotification {
+    static let keyboardShowNotification = NotificationDescriptor<KeyboardData>(name: Notification.Name.UIKeyboardWillShow, convert: KeyboardData.init)
+    static let keyboardHiderNotification = NotificationDescriptor<KeyboardData>(name: Notification.Name.UIKeyboardWillHide, convert: KeyboardData.init)
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: remove 'r' from keyboardHiderNotification

File: _posts/2017-08-28-typed-notification-ios.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+}
+{% endhighlight %}
+
+We have also encapsulated these properties in `SystemNotification` struct, so that our global namespace doesn't get polluted with all different types of notification descriptors.
+
+### UIViewController usage
+
+As we have these properties already define, using in a UIViewController will become very simple and we don't have to track the notification for deregister.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: already defined, using them in a .....

File: _posts/2017-08-28-typed-notification-ios.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+private var keyboardShowToken: Token?
+private var keyboardHideToken: Token?
+
+override func viewDidLoad() {
+    super.viewDidLoad()
+
+    registerForKeyboardNotifications()
+}
+
+func registerForKeyboardNotifications() {
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: Mark this function for better understanding.

File: _posts/2017-08-28-typed-notification-ios.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+        keyboardHideToken = NotificationCenter.default.addObserver(descriptor: SystemNotification.keyboardHideNotification) { keyboardData in
+            //Custom logic related to view only
+        }
+    }
+{% endhighlight %}
+
+This way we are also freeing up our UIViewController from other logic and only allowing it to handle views. This is also highly reusable as all the repetitive code is moved to a single place and this makes our code `DRY`.
+
+For MVVM architecture above implementation fits perfect, as in MVVM a viewController should only handle view related logic only.
```
Comment (unleashedspeed) [inline _posts/2017-08-28-typed-notification-ios.md]: we can make MVVM a hyperlink here and provide a link which leads to a good article about MVVM

**Review by unleashedspeed — CHANGES_REQUESTED** [review]

**Comment by aksswami** [comment]
Thanks @unleashedspeed, I have resolved all the comments as per your suggestion.

**Review by unleashedspeed — APPROVED** [review]

**Review by anujmiddha — APPROVED** [review]

**Pull request merged**